File: 프로그래머스/L3/등대/등대.py
import sys
sys.setrecursionlimit(10 ** 9)
def solution(n, lighthouse):
    answer = 0
    graph = [[] for i in range(n+1)]
    visited = [False for i in range(n+1)]
    
    for a,b in lighthouse:
        graph[a].append(b);
        graph[b].append(a);
    
    def dfs(idx):
        visited[idx] = True
        p,np = 1,0
        for nIdx in graph[idx]:
            if visited[nIdx]:
                continue
            resP, resNp = dfs(nIdx)
            p += min(resP,resNp)
            np += resP
        return p,np
    return min(dfs(1))

File: 프로그래머스/L3/등대/등대.js
function solution(n, lighthouse) {
  const graph = new Array(n + 1).fill(null).map((_) => new Array());
  const visited = new Array(n + 1).fill(false);
  for (let [a, b] of lighthouse) {
    graph[a].push(b);
    graph[b].push(a);
  }
  const dfs = (idx) => {
    visited[idx] = true;
    let p = 1;
    let np = 0;
    for (let nIdx of graph[idx]) {
      if (visited[nIdx]) continue;
      const [resP, resNp] = dfs(nIdx);
      p += Math.min(resP, resNp);
      np += resP;
    }
    return [p, np];
  };
  return Math.min(...dfs(1));
}

const res = solution(9, [
  [1, 2],
  [2, 3],
  [3, 4],
  [5, 6],
  [6, 7],
  [7, 8],
  [8, 9],
]);

console.log(res);
